salt/states/module.py

```
@@ -658,31 +658,29 @@ def _legacy_run(name, **kwargs):
         if kwargs["returner"] in returners:
             returners[kwargs["returner"]](ret_ret)
     ret["comment"] = f"Module function {name} executed"
-    ret["result"] = _get_result(mret, ret["changes"])
+    ret["result"] = _get_result(mret)
 
     return ret
 
 
-def _get_result(func_ret, changes):
+def _get_result(func_ret):
     res = True
     # if mret is a dict and there is retcode and its non-zero
     if isinstance(func_ret, dict) and func_ret.get("retcode", 0) != 0:
         res = False
         # if its a boolean, return that as the result
     elif isinstance(func_ret, bool):
         res = func_ret
-    else:
-        changes_ret = changes.get("ret", {})
-        if isinstance(changes_ret, dict):
-            if isinstance(changes_ret.get("result", {}), bool):
-                res = changes_ret.get("result", {})
-            elif changes_ret.get("retcode", 0) != 0:
-                res = False
-            # Explore dict in depth to determine if there is a
-            # 'result' key set to False which sets the global
-            # state result.
-            else:
-                res = _get_dict_result(changes_ret)
+    elif isinstance(func_ret, dict):
```

There is no longer an else case for when `func_ret` is not a `bool` nor a `dict`, the return will be `True`. The `func_ret` not being a `bool` nor `dict` is unexpected so should throw an exception or at least return `False`. Also we are checking `isinstance(func_ret, dict)` multiple times (also on line 669). We should just check once and have the `func_ret.get("retcode", 0) != 0` check be inside it.

---

There was not really a plain else case before, as the `else` had the `if isinstance(changes_ret, dict)` inside which itself had no else case.
And I think it is ok or even wanted to have it return `True` in such cases: Not every execution module function does give a clear successful/unsuccessful indication.

---

> Not every execution module function does give a clear successful/unsuccessful indication.
In that case I would expect `func_ret` to be `None` which I recommend we handle in the way you deem correct, but an unexpected data type should either return the truth/`bool` value of it self (i.e. `bool(func_ret)` so empty list `[]` is False and list with value is True) or should throw an exception because we are getting an unexpected value. We should not do the worst case though of returning `True` that function was successful it wasn't and returned a bad value.

---

Well, there seems to be no hard standard for the return of execution modules. Unlike state modules.
Returning `None` or `[]` could simply mean that it had nothing to do.
For example `saltutil.sync_modules` does return a list of synced modules. But if there was nothing to sync, the list is empty, which is totally fine and actually a success.

However, there are some typical ways how some module functions indicate the succession: By a returning a bool or by dict keys `result` or `retcode`. Everything else would be a guess.
Using module.run is just some kind of a workaround anyway. If a proper determination of the result is needed, one should look for a state module or make one.

Also this PR is just meant to fix the limited (and often sufficient) way the function already had to determine the succession. Not to redesign it.